permission-ai (#1)
uma-victor1/fine-grained-permission: project/docs.md

> # Fine-Grained Security for AI Agents

Reviewer: ```suggestion
# Fine-Grained Permissions for AI Agents
```

> This example demonstrates implementing a secure financial advisor agent using PydanticAI with four distinct security perimeters. It shows how to build a production-ready AI system that enforces certification requirements, regulatory compliance, and proper risk management.

Reviewer: ```suggestion
This example demonstrates implementing a secure financial advisor agent using PydanticAI with four distinct access control perimeters. It shows how to build a production-ready AI system that enforces certification requirements, regulatory compliance, and proper risk management.
```

Demonstrates:

- [dynamic system prompt](../agents.md#system-prompts)
- [structured `result_type`](../results.md#structured-result-validation)
- [result validation](../results.md#result-validators-functions)
- [tools](../tools.md)

> ## Security Perimeters Overview

Reviewer: ```suggestion
## Access Control Perimeters Overview
```

> ### Prompt Validation

Reviewer: Please see the new names, and replace it anywhere needed:

1. Prompt Filtering
2. Data Protection / RAG Data Protection
3. Secure External Access
4. Response Enforcement

> Validates whether users have explicitly consented to receive AI-generated financial advice.

Reviewer: In *every* perimeter, before explaining the financial advisor example, give the general context of the perimeter. For example: The first perimeter to secure, is filtering the Prompt per the user permissions.......

- Document classification levels (public, restricted, confidential)
- User clearance level verification
- Regulatory compliance for information access
- Audit trail of document access

```python
@financial_agent.tool
async def access_financial_knowledge(
ctx: RunContext[PermitDeps],
query: FinancialQuery
) -> List[str]:

```

### Action Permissions

Controls permissions for sensitive financial operations, particularly portfolio modifications.

- Portfolio ownership verification
- User authorization level
- Transaction limits compliance
- Account access restrictions

```python
@financial_agent.tool
async def check_action_permissions(
ctx: RunContext[PermitDeps],
action: str,
context: UserContext
) -> bool:

```

### Response Validation

Ensures all financial advice responses meet regulatory requirements and include
necessary disclaimers.

- Automated advice detection using Permit
- Regulatory disclaimer insertion
- Compliance verification

```
@financial_agent.tool
async def validate_financial_response(
ctx: RunContext[PermitDeps],
response: FinancialResponse,
context: UserContext
) -> Dict[str, bool]:

```

## Running the Example

> First, configure your Permit.io environment:

Reviewer: ```suggestion
First, configure your Permit.io environment (you can get a [free API key here](https://app.permit.io)):
```

```bash
export PERMIT_API_KEY='your-api-key'
```

Update the API key and PDP url in the example code:

```python
Permit(

> token="YOUR_API_KEY", # Replace with your actual API key

Reviewer: It shoulde be the environment varialbe

pdp="http://localhost:7766",
)
```

> Here's a visualization of the control flow:

Reviewer: After you'll creating the whole policy in Permit, I recommend to put here a screenshot (or in text) of the policy table, so people can understand the polciies

```
User Request
     ↓
PydanticAI Framework
     ↓
validate_financial_query  →  [If fails, stops here]
     ↓
advisor certification check
     ↓
Claude AI Processing     ←  [PydanticAI hands control to Claude]
     ↓
validate_financial_advice →  [If fails, stops here]
     ↓
access_financial_knowledge →  [If fails, stops here]
     ↓
Compliance Check
     ↓
Response to User
```

## Running the Example

With [dependencies installed and environment variables set](./index.md#usage), run:

```bash

> python/uv-run -m pydantic_ai_examples.secure-ai-agent

Reviewer: I think we are missing the point here of explaining people to run the config.py for the setup
